elfpy/utils/apeworx_integrations.py

```diff
@@ -707,43 +724,56 @@ def attempt_txn(
     """
+    # allow inconsistent return, since we throw an error if the transaction fails after all attempts
+    # pylint: disable=inconsistent-return-statements
+
     mult = kwargs.pop("mult") if hasattr(kwargs, "mult") else 2
     priority_fee_multiple = kwargs.pop("priority_fee_multiple") if hasattr(kwargs, "priority_fee_multiple") else 5
     if isinstance(contract_txn, ContractTransactionHandler):
         abi, args = select_abi(method=contract_txn, args=args)
         contract_txn = ContractTransaction(abi=abi, address=contract_txn.contract.address)
-    latest = agent.provider.get_block("latest")
-    if latest is None:
-        raise ValueError("latest block not found")
-    if not hasattr(latest, "base_fee"):
-        raise ValueError("latest block does not have base_fee")
-    base_fee = getattr(latest, "base_fee")
 
     # begin attempts, indexing attempt from 1 to mult (for the sake of easy calculation)
     for attempt in range(1, mult + 1):
-        kwargs["max_fee_per_gas"] = int(base_fee * attempt + agent.provider.priority_fee * priority_fee_multiple)
+        latest = agent.provider.get_block("latest")
+        if latest is None:
+            raise ValueError("latest block not found")
+        if not hasattr(latest, "base_fee"):
+            raise ValueError("latest block does not have base_fee")
+        base_fee = getattr(latest, "base_fee")
+        log_and_show(f"latest block {fmt(getattr(latest, 'number'))} has base_fee {base_fee/1e9:,.3f}")
+
         kwargs["max_priority_fee_per_gas"] = int(
-            agent.provider.priority_fee * priority_fee_multiple + base_fee * (attempt - 1)
-        )
+            agent.provider.priority_fee * priority_fee_multiple * attempt
+        )  # agent.provider.priority_fee * priority_fee_multiple + base_fee * (attempt - 1)
+        kwargs["max_fee_per_gas"] = int(base_fee * (1 + attempt)) + kwargs["max_priority_fee_per_gas"]
         kwargs["sender"] = agent.address
         kwargs["nonce"] = agent.provider.get_nonce(agent.address)
         kwargs["gas_limit"] = 1_000_000
         # if you want a "STATIC" transaction type, uncomment the following line
         # kwargs["gas_price"] = kwargs["max_fee_per_gas"]
-        log_and_show(f"txn attempt {attempt} of {mult} with {kwargs=}")
+        formatted_items = []
+        for k, v in kwargs.items():
+            value = fmt(v / 1e9) if "fee" in k else fmt(v)
+            formatted_items.append(f"{k}={value}")
+        log_and_show(f"txn attempt {attempt} of {mult} with {', '.join(formatted_items)}")
         serial_txn: TransactionAPI = contract_txn.serialize_transaction(*args, **kwargs)
         prepped_txn: TransactionAPI = agent.prepare_transaction(serial_txn)
         signed_txn: Optional[TransactionAPI] = agent.sign_transaction(prepped_txn)
-        log_and_show(f" => sending {signed_txn=}")
+        logging.debug(" => sending signed_txn %s", signed_txn)
```

Comment:
What is the reasoning between when you want to "log and show" and when you want to use normal logging? I'm wondering if we should just set up a logging config parameter that supports the "log and show" method. The way you have it now, it is up to the programmer to be opinionated about when it is appropriate to do one or the other, which I don't think is the right model with logging. Let's discuss later and make an issue for it if we agree.

Reply:
it's just my ongoing obsession with seeing things on my console, and hating using log files.

don't want to create overhead for the coder. I tried to make this purely optional, by adding an extra handler to our existing logging setup. but committing it here feels like everyone reviewing this code is going to wonder when to use `log_and_show` and when to use regular logging. so I'm kinda imposing my preference on them, through that. feels like there should be a better way. maybe a... "print every error to stdout" as an additional toggle in our logging?

Reply:
I guess the current thinking is "do I want to see this on every run, or only when debugging?"